pkg/agent/sysadvisor/plugin/qosaware/resource/memory/plugin/provisioner/policy_canonical.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2022 The Katalyst Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provisioner
+
+import (
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/plugin/qosaware/resource/helper"
+	"github.com/kubewharf/katalyst-core/pkg/consts"
+	"github.com/kubewharf/katalyst-core/pkg/util/metric"
+
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/metacache"
+	"github.com/kubewharf/katalyst-core/pkg/config"
+	"github.com/kubewharf/katalyst-core/pkg/metaserver"
+	"github.com/kubewharf/katalyst-core/pkg/metrics"
+	"github.com/kubewharf/katalyst-core/pkg/util/general"
+	"github.com/kubewharf/katalyst-core/pkg/util/machine"
+)
+
+type PolicyCanonical struct {
+	*PolicyBase
+	memoryProvisions machine.MemoryDetails
+}
+
+func (p *PolicyCanonical) Update() error {
+	var (
+		memFreeNuma  metric.MetricData
+		memTotalNuma metric.MetricData
+	)
+
+	memoryProvisions := make(machine.MemoryDetails)
+	memoryTotals := make(machine.MemoryDetails)
+
+	// 1. get all numa nodes
+	allNUMAs := p.metaServer.CPUDetails.NUMANodes()
+
+	// 2. get all reclaimed cores containers
+	availNUMAs, reclaimedCoresContainers, err := helper.GetAvailableNUMAsAndReclaimedCores(p.conf, p.metaReader, p.metaServer)
+
+	if err != nil {
+		return err
+	}
+
+	// 3. get all numa memory provision
+	availNUMATotal := float64(0)
+	for _, numaID := range availNUMAs.ToSliceInt() {
+		memFreeNuma, err = p.metaServer.GetNumaMetric(numaID, consts.MetricMemFreeNuma)
+		if err != nil {
+			return err
+		}
+
+		memTotalNuma, err = p.metaServer.GetNumaMetric(numaID, consts.MetricMemTotalNuma)
+		if err != nil {
+			return err
+		}
+
+		memoryProvisions[numaID] += uint64(memFreeNuma.Value)
+		memoryTotals[numaID] += uint64(memTotalNuma.Value)
+		availNUMATotal += memTotalNuma.Value
+		general.InfoS("numa memory free",
+			"numaID", numaID,
+			"numaFree", general.FormatMemoryQuantity(memFreeNuma.Value))
+	}
+
+	// 4. add reclaimed cores containers memory
+	for _, containerInfo := range reclaimedCoresContainers {
+		// if nil, we should range all numa nodes
+		if containerInfo.TopologyAwareAssignments == nil {
+			c := uint64(containerInfo.MemoryRequest)
+			c /= uint64(len(memoryProvisions))
```

Comment:
if no available numa, it may panic here?

Reply:
done

Reply:
I did a no available numa check before this logic.